nn/conformer.py

```diff
@@ -0,0 +1,217 @@
+"""
+Conformer code.
+Ref: https://arxiv.org/abs/2005.08100
+"""
+
+from typing import Tuple, List, Union
+from .. import nn
+from . import LayerRef
+
+
+class _PositionwiseFeedForward(nn.Module):
+  """
+  Conformer position-wise feedforward neural network layer
+      FF -> Activation -> Dropout -> FF
+  """
+
+  def __init__(self, dim_model: int, dim_ff: int, dropout: float, activation, l2: float = 0.0):
+    """
+    :param dim_model:
+    :param dim_ff:
+    :param dropout:
+    :param activation:
+    :param l2:
+    """
+    super().__init__()
+
+    self.dropout = dropout
+    self.activation = activation
+
+    self.linear1 = nn.Linear(n_out=dim_ff, l2=l2)
+    self.linear2 = nn.Linear(n_out=dim_model, l2=l2)
+
+  def forward(self, inp: LayerRef) -> LayerRef:
+    return self.linear2(nn.dropout(self.activation(self.linear1(inp)), dropout=self.dropout))
+
+
+class _ConformerConvBlock(nn.Module):
+  """
+  Conformer convolution block
+      FF -> GLU -> depthwise conv -> BN -> Swish -> FF
+  """
+
+  def __init__(self, dim_model: int, kernel_size: Tuple[int], l2: float = 0.0, batch_norm_eps: float = 1e-5,
+      batch_norm_momentum: float = 0.1, batch_norm_update_sample_only_in_training=True,
+      batch_norm_delay_sample_update=True, batch_norm_other_opts=None):
+    """
+    :param dim_model:
+    :param kernel_size:
+    :param l2:
+    """
+    super().__init__()
+
+    self.positionwise_conv1 = nn.Linear(n_out=dim_model * 2, l2=l2)
+    self.depthwise_conv = nn.Conv(n_out=dim_model, filter_size=kernel_size, groups=dim_model, l2=l2, padding='same')
+    self.positionwise_conv2 = nn.Linear(n_out=dim_model, l2=l2)
+
+    if batch_norm_other_opts is None:
+      batch_norm_other_opts = {}
+
+    self.batch_norm = nn.BatchNorm(epsilon=batch_norm_eps, momentum=batch_norm_momentum,
```

Comment:
If you anyway need to continue the line with a function call, do not add any arguments in this line but only in the following lines.
```suggestion
    self.batch_norm = nn.BatchNorm(
      epsilon=batch_norm_eps, momentum=batch_norm_momentum,
```